e2e/nomostest/watch.go

```diff
@@ -88,166 +95,230 @@ func WatchObject(nt *NT, gvk schema.GroupVersionKind, name, namespace string, pr
+	// Since items are typed, we have to use some reflection to extract them.
+	// But since we filtered by name and namespace, there should only be one item.
+	cObj, err := getObjectFromList(cObjList, name, namespace)
 	if err != nil {
-		return errors.Wrapf(err, "%s unable to understand list result %#v",
-			errPrefix, cObjList)
+		return errors.Wrap(err, errPrefix)
 	}
+
 	// Save the predicate errors from the last state change and return them
 	// if the watch closes before the predicates all pass.
 	var evalErrs []error
-	// Iterate through the list to find the desired object, if it exists
+	// Save the last known object state for diffing with the new state.
 	var prevObj client.Object
-	for _, rObj := range items {
-		cObj, ok := rObj.(client.Object)
-		if !ok {
-			return errors.Wrapf(
-				WrongTypeErr(rObj, client.Object(nil)),
-				"%s unexpected list item type", errPrefix)
-		}
-		if cObj.GetName() != name {
-			// Ignore other objects, if any.
-			// Should never happen, due to name selector.
-			continue
-		}
-		prevObj = cObj
-		break
-	}
-	if prevObj == nil {
+	if cObj == nil {
+		// Normally, we ignore NotFound and wait for the object to exist,
+		// unless the caller is specifically waiting for NotFound.
+		// TODO: Refactor Predicates to accept nil objects so we can remove this hacky workaround.
 		if spec.errorOnNotFound {
 			return &WatchObjectNotFoundError{
 				GroupVersionKind: gvk,
 				Name:             name,
 				Namespace:        namespace,
 			}
 		}
-		nt.DebugLogf("%s GET Not Found", errPrefix)
+		nt.T.Logf("%s GET Not Found", errPrefix)
 	} else {
+		nt.T.Logf("%s GET Found", errPrefix)
 		// Log the initial/current state as a diff to make it easy to compare with update diffs.
 		// Use diff.Diff because it doesn't truncate like cmp.Diff does.
 		// Use log.AsYAMLWithScheme to get the full scheme-consistent YAML with GVK.
 		nt.DebugLogf("%s GET Diff (+ Current):\n%s",
-			errPrefix, log.AsYAMLDiffWithScheme(nil, prevObj, nt.scheme))
-		// Evaluate predicates
-		evalErrs = eval(prevObj)
+			errPrefix, log.AsYAMLDiffWithScheme(nil, cObj, nt.scheme))
+		evalErrs = evaluate(cObj, predicates)
 		if len(evalErrs) == 0 {
 			// Success! All predicates passed!
 			return nil
 		}
+		prevObj = cObj
 	}
 
-	rv := cObjList.GetResourceVersion()
+	// Specify ResourceVersion to ensure the Watch starts where the List stopped.
+	initialResourceVersion := cObjList.GetResourceVersion()
 
-	// Create a new empty list to use with WATCH
-	cObjList, err = newObjectListForObjectKind(gvk, nt.scheme)
-	if err != nil {
-		return err
-	}
+	// Don't need to specify RV in the options. watchtools.Until does it for us.
+	// listOpts.ResourceVersion = cObjList.GetResourceVersion()
 
-	// Specify ResourceVersion to ensure the Watch starts where the List stopped.
-	cObjList.SetResourceVersion(rv)
+	// ResourceVersionMatch is forbidden for WATCH.
+	// listOpts.ResourceVersionMatch = metav1.ResourceVersionMatchNotOlderThan
```

Comment:
did you intend to leave the commented code here and above?

Reply:
Removed.